### PR TITLE
Issue #826: Support AssertJ/Google Truth fail() methods in RequireFailForTryCatchInJunitCheck

### PR DESCRIPTION
**Pull request by sebthom**
Solves #826

### CODE REVIEW

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java [hunk 1]
```diff
@@ -84,35 +91,100 @@
...
+     * Fully qualified identifier of static methods whose call would satisfies this check.
      */
-    private boolean importTest;
+    private static final Set<String> FAIL_METHODS = new HashSet<>(Arrays.asList(
+            "com.google.common.truth.Truth#assert_.fail",
+            "org.assertj.core.api.Assertions#fail",
+            "org.assertj.core.api.Assertions#failBecauseExceptionWasNotThrown",
+            "org.junit.Assert#fail",
+            "org.junit.jupiter.api.Assertions#fail"));
+
```
Comment (pbludov) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:104]: May be this should be extracted to some property?

**Comment by pbludov** [comment]
@sebthom please make CI happy.

**Comment by sebthom** [comment]
@pbludov Thanks for reviewing. I fixed the CI issues as far as I could, now only the following fail and I don't see that I could do anything about it:

1. https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/750731706
   This fails with "mkdir: cannot create directory ‘net.sf.eclipsecs.doc/docs’: File exists" - I don't think this is related to the PR
1. https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/750731709
   Here PMD complains about code coverage, however the code coverage of the **RequireFailForTryCatchInJunitCheck** class is 98,7% - so I again don't think this is related to the PR
1. https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/750731713
   Fails because the PR is for a not yet approved issue

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java (hunk 1 above)
Comment (romani) [reply sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:105]: @sebthom , please update javadoc with new methods to detect.
javadoc is the only way for user to read what Check is doing.

Comment (romani) [reply sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:105]: we can skip new property for now,  but if somebody would need new method - we have to.

**Comment by romani** [comment]
problem is test coverage - https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/750731709#L793
please add more tests.
```
[INFO] --- jacoco-maven-plugin:0.7.1.201405082137:check (default-check) @ sevntu-checks ---

[INFO] Analyzed bundle 'sevntu-checks' with 80 classes

[WARNING] Rule violated for class 
com.github.sevntu.checkstyle.checks.coding.RequireFailForTryCatchInJunitCheck: 
branches covered ratio is 0.96, but expected minimum is 0.98
```

**Comment by sebthom** [comment]
> 
> problem is test coverage - [travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/750731709#L793](https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/750731709#L793)

That makes no sense, before my PR the test coverage of `RequireFailForTryCatchInJunitCheck` - which is the only class I modified - was **97,9%**:
image

After my PR the coverage increased to **98,6%**:
image

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java (hunk 1 above)
Comment (sebthom) [reply sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:113]: > @sebthom , please update javadoc with new methods to detect.
> javadoc is the only way for user to read what Check is doing.
@romani I extended the javadoc as requested

**Comment by romani** [comment]
@sebthom , do you mean jacoco has problems with limits calculations ?

**Comment by romani** [comment]
@sebthom , please fix LineLengh, to let me see one more time  what is wrong with limits, it use to be very reliable way to check of coverage.

**Comment by sebthom** [comment]
> @sebthom , please fix LineLengh

@romani I fixed it. Interestingly I already had the eclipsecs plugin activated with the sevntu-checks.xml rule set but it did not report the LineLength violation.

**Comment by sebthom** [comment]
> @sebthom , do you mean jacoco has problems with limits calculations ?

I don't really know what is going on in the CI and how the coverage is calculated. But generally I think for PRs it should not be an absolute value, but rather a difference check between before/after applying the PR.

**Comment by romani** [comment]
on local I have bunch of problems like:
```
~/java/github/sevntu-checkstyle/sevntu.checkstyle [sebthom/master L|✔] 
$ ./.ci/travis.sh sevntu-checks


[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.4:jar (attach-javadocs) on project sevntu-checks: MavenReportException: Error while generating Javadoc: 
[ERROR] Exit code: 1 - 
sevntu-checkstyle/sevntu.checkstyle/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:56: 
error: invalid uri: "https://truth.dev/api/latest/com/google/common/truth/StandardSubjectBuilder.html#fa
[ERROR]  * <li><a href="https://truth.dev/api/latest/com/google/common/truth/StandardSubjectBuilder.html#fa
[ERROR]           ^
[ERROR]   il()"

```

**Comment by sebthom** [comment]
> on local I have bunch of problems like:
> 
> ```
> ~/java/github/sevntu-checkstyle/sevntu.checkstyle [sebthom/master L|✔] 
> $ ./.ci/travis.sh sevntu-checks
> 
> 
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.4:jar (attach-javadocs) on project sevntu-checks: MavenReportException: Error while generating Javadoc: 
> [ERROR] Exit code: 1 - 
> sevntu-checkstyle/sevntu.checkstyle/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:56: 
> error: invalid uri: "https://truth.dev/api/latest/com/google/common/truth/StandardSubjectBuilder.html#fa
> [ERROR]  * <li><a href="https://truth.dev/api/latest/com/google/common/truth/StandardSubjectBuilder.html#fa
> [ERROR]           ^
> [ERROR]   il()"
> ```

Hmm ok, I need your advice here, how can I satisfy the line length check in the JavaDoc if the URLs are longer than the limit?

**Comment by romani** [comment]
>how can I satisfy the line length check in the JavaDoc 

There should be specific suppress on violation over url. Please make try to keep only url on line. I will review why it is not suppressed

**Comment by sebthom** [comment]
@romani I think travis ran out of credits. maybe time to migrate to github actions.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/9fa785ef20536f6c9482d5ae6d09f2d31fb44637/config/checkstyle_checks.xml#L214

This supposed to cover such cases

**Comment by sebthom** [comment]
@romani ok, I am stuck with the LineLength violations. I don't know how to properly suppress them in the project. I could just remove the URLs from the JavaDoc but I feel like they are valuable information in that context.

**Comment by romani** [comment]
I will help

**Comment by romani** [comment]
fix is
```
~/java/github/sevntu-checkstyle/sevntu.checkstyle/sevntu-checks [sebthom/master L|✚ 1] 
05:28 $ git diff
diff --git a/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java b/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java
index eaca844..b1fff05 100644
--- a/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java
+++ b/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java
@@ -53,17 +53,22 @@ import com.puppycrawl.tools.checkstyle.utils.AnnotationUtil;
  * by a test method. It must contain the annotation.
  * Failures are identified by a method call to either:
  * <ul>
- * <li><a href="https://truth.dev/api/latest/com/google/common/truth/StandardSubjectBuilder.html#fail()">
+ * <li>
+ * <a href="https://truth.dev/api/latest/com/google/common/truth/StandardSubjectBuilder.html#fail()">
  * com.google.common.truth.Truth#assert_.fail</a>
  * <li><a href="http://junit.sourceforge.net/junit3.8.1/javadoc/junit/framework/Assert.html#fail()">
  * junit.framework.Assert#fail</a>
- * <li><a href="https://www.javadoc.io/static/org.assertj/assertj-core/3.19.0/org/assertj/core/api/Assertions.html#fail(java.lang.String)">
+ * <li>
+ * <a href="https://www.javadoc.io/static/org.assertj/assertj-core/3.19.0/org/assertj/core/api/Assertions.html#fail(java.lang.String)">
  * org.assertj.core.api.Assertions#fail</a>
- * <li><a href="https://www.javadoc.io/static/org.assertj/assertj-core/3.19.0/org/assertj/core/api/Assertions.html#failBecauseExceptionWasNotThrown(java.lang.Class)">
+ * <li>
+ * <a href="https://www.javadoc.io/static/org.assertj/assertj-core/3.19.0/org/assertj/core/api/Assertions.html#failBecauseExceptionWasNotThrown(java.lang.Class)">
  * org.assertj.core.api.Assertions#failBecauseExceptionWasNotThrown</a>
- * <li><a href="https://junit.org/junit4/javadoc/latest/org/junit/Assert.html#fail()">
+ * <li>
+ * <a href="https://junit.org/junit4/javadoc/latest/org/junit/Assert.html#fail()">
  * org.junit.Assert#fail</a>
- * <li><a href="https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#fail()">
+ * <li>
+ * <a href="https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#fail()">
  * org.junit.jupiter.api.Assertions#fail</a>
  * </ul>
  * </p>
```

but there rabase to latest our master to resolve other violations and there are still problems with javadoc over html tags, please continue investigation.

**Comment by sebthom** [comment]
@romani I applied your changes and rebased

**Comment by romani** [comment]
Travis demands more coverage
```
[WARNING] Rule violated for class com.github.sevntu.checkstyle.checks.coding.RequireFailForTryCatchInJunitCheck: branches covered ratio is 0.97, but expected minimum is 0.98
```

**Comment by sebthom** [comment]
@romani the RequireFailForTryCatchInJunitCheck now already has the same coverage as before the PR. The **only** branch not covered is the one in the following screenshot which also was not covered before my PR. Also I don't see how to cover this in a *meaningful* way unless. I really don't understand what Travis is calculating there.
image

**Comment by romani** [comment]
Do extra test method like https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/827#issuecomment-775660097

**Comment by sebthom** [comment]
Actually there already exists a test case that does this in RequireFailForTryCatchInJunitCheckTest. But jacoco does not seem to track it correctly. Looks like a bug in jacoco to me because when I step-debug through the test case execution I can actually see that all statements are invoked.

image

**Comment by sebthom** [comment]
I still think there is something fundamentally wrong in the coverage calculation. why does one have to cover more code than one changed to pass the coverage test??

When I run jacoco in Eclipse I get a coverage of 98.9%. So I have no idea why travis talks about 0.97%

image

**Comment by romani** [comment]
we have very simeple rule, coverage numbers do not go down, they stay same  or increase.

**Comment by romani** [comment]
@sebthom , please update config to have 97 as limit.

I confirmed on local that coverage is actually 100% (by IDEA coverage tool), default is covered for sure, I do not know why it is a probem. Your update does not make this problem, it existing in jacoco report even in master branch.

**Comment by romani** [comment]
Squash commits in one, CI should pass

**Review by pbludov — APPROVED** [review]

**Comment by sebthom** [comment]
Now another travis job fails with "The version of Java (1.8.0_252) you have used to run this analysis is deprecated and we stopped accepting it. Please update to at least Java 11."

**Comment by romani** [comment]
https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/758953591#L595 CI demand message to be in correct form. make it single line.

**Comment by sebthom** [comment]
@romani I squashed the commits. now another travis job that succeeded before fails again with LineLength violations,  even though I didn't change anything else...the [fix](https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/827#issuecomment-776707207) you suggested before is still in place. so I am out if idea what's the problem now.

**Comment by romani** [comment]
weird, I have no violations on my local with your branch ...
but I can not merge because it will expose problem to all other PRs.

**Comment by romani** [comment]
I reproduced issue by execution 
```
~/java/github/sevntu-checkstyle/sevntu.checkstyle [sebthom/master L|✔] 
$ ./.ci/travis.sh sevntu-checks
....
[INFO] --- maven-checkstyle-plugin:3.1.1:check (default) @ sevntu-checks ---
[INFO] There are 4 errors reported by Checkstyle 8.29 with sevntu-checks.xml ruleset.
[ERROR] src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:[58] (extension) LineLengthExtended: Line is longer than 100 characters. Current line length is 101.
[ERROR] src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:[64] (extension) LineLengthExtended: Line is longer than 100 characters. Current line length is 135.
[ERROR] src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:[67] (extension) LineLengthExtended: Line is longer than 100 characters. Current line length is 162.
[ERROR] src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java:[73] (extension) LineLengthExtended: Line is longer than 100 characters. Current line length is 122.

```

**Comment by romani** [comment]
@sebthom , I merged fix for this problem - https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/829

please rebase one more time, CI should be green after rebase.

**Pull request merged**